### ISSUE TITLE
monetization "stopped" state and monetizationstop / monetizationpending events

### ISSUE THREAD

**Issue by adrianhopebailie** [interledger/webmonetization.org#25]
From web-monetization created by [sublimator](https://github.com/sublimator): adrianhopebailie/web-monetization#24  

@marcoscaceres 
For the moment using terms inline with the existing implementation :) 

Having done some (non deployed) experiments internally re: pausing of streams and associated states/events, putting out this to get some feedback. 

Adding a `stopped` state and redefining `pending`:
```typescript
/**
 * Streaming has been initiated, yet first non zero packet is "pending"
 * It will normally transition from this `state` to `started`, yet not
 * always.
 */
export type PendingState = 'pending'

/**
 * Streaming has received a non zero packet and is still active
 */
export type StartedState = 'started'

/**
 * Streaming is inactive.
 * In the interest of keeping the amount of states necessary to handle down
 * this could mean a variety of things:
 *  - monetization may not be supported (no provider enabled)
 *  - may not have started yet
 *  - paused (potentially resumed)
 *  - finished completely (awaiting another request)
 *  - payment request denied by user intervention
 */
export type StoppedState = 'stopped'
```

`document.monetization` object

```typescript
export type MonetizationState = PendingState | StartedState | StoppedState
export type MonetizationObject = EventTarget & {
  state: MonetizationState
}
```

Adding `monetizationstop`/`monetizationpending` events:
```typescript
export interface MonetizationCommonEventDetail {
  // Web-Monetization-Id header present in the SPSP request
  requestId: string
  // The meta[@name="monetization"] @content value
  paymentPointer: string
}

export interface MonetizationStartEvent {
  type: 'monetizationstart'
  detail: MonetizationCommonEventDetail
}

export interface MonetizationProgressEvent {
  type: 'monetizationprogress'
  detail: {
    // The amount * received * at the destination specified in the SPSP endpoint
    amount: string
    assetCode: string
    assetScale: number
  } & MonetizationCommonEventDetail
}

export interface MonetizationPendingEvent {
  type: 'monetizationpending'
  detail: MonetizationCommonEventDetail
}

export interface MonetizationStopEvent {
  type: 'monetizationstop'
  detail: MonetizationCommonEventDetail
}
```

Excuse the TypeScript vs Web-IDL

**Comment by adrianhopebailie** [comment]
The current impl events are essentially CustomEvent(...) with `detail`, not quite inline with the spec here. I had a quick look at the WEB streams API and it's not apparent to me (from that link at least) how you would use that for events. Would you JSON parse the chunks as they came in?

**Comment by adrianhopebailie** [comment]
Thanks @sublimator - the TypeScript is quite helpful! 

I think we need to do also do a full comparison of what this could look like using a `ReadableStream` http://streams.spec.whatwg.org ... then we can also evaluate if we need to special-case anything (hopefully `ReadableStream` can capture everything we need). 

Is that something might have time to look into?

**Comment by adrianhopebailie** [comment]
I am not really familiar with streams, what's the advantage there?
Last I looked it was for binary data ? 
Having a quick squizz now

**Comment by adrianhopebailie** [comment]
Or are you talking about using the streams for the actual STREAM websocket connections?

**Comment by adrianhopebailie** [comment]
> I am not really familiar with streams, what's the advantage there?

All the work is done for us. Web platform primitive with well understood semantics. 

> Last I looked it was for binary data ?

No. They are for anything that "streams". 

> Having a quick squizz now

🤘

> Would you JSON parse the chunks as they came in?

We can separate the JSON part out... it's more about "data" rather. Consider this in terms of a "session":

 * Money "starts" streaming ...
 * ... more money steams as a particular rate... chunks are processed/acknowledged/transformed/displayed/etc.
 * Stream is eventually "closed" (e.g., money runs out, payment is complete, etc.). 

> Or are you talking about using the streams for the actual STREAM websocket connections?

Nope... don't know anything about STREAM - but maybe they work together?

**Comment by adrianhopebailie** [comment]
@marcoscaceres 
Streams are quite interesting. Definitely food for thought. Streams have a definite 'finality' to them that the above state representation doesn't. As you can see, for better or worse, the "stopped" state is quite ambiguous above. How did you imagine a user getting a handle on a stream ? Would it be a property of an event ? Would you have more than one ReadableStream for a given "session" (defined by requestId above) ? start / stop of a stream spanning a start / pause ? **Should a new requestId be generated upon restarting/resuming a stream?**

(edit: avoiding another email notification)
I do wonder about multiple consumers of a given stream as well. Would each consumer get its own handle ?

I note that Streams aren't super well supported yet:  
https://developer.mozilla.org/en-US/docs/Web/API/WritableStream#Browser_Compatibility
https://developer.mozilla.org/en-US/docs/Web/API/ReadableStream#Browser_Compatibility

**Comment by adrianhopebailie** [comment]
Note I don't have a concrete proposal yet about how this should work... just feels like it's worth exploring. I'll hopefully get a bit of time to look into this in detail in the next few weeks, but maybe someone closer to the spec can do a deep dive? @tobie maybe?

**Comment by adrianhopebailie** [comment]
> just feels like it's worth exploring

Noted :)

**Comment by adrianhopebailie** [comment]
```typescript
/**
 * Streaming is inactive.
 * In the interest of keeping the amount of states necessary to handle down
 * this could mean a variety of things:
 *  - monetization may not be supported (no provider enabled)
 *  - may not have started yet
 *  - paused (potentially resumed)
 *  - finished completely (awaiting another request)
 *  - payment request denied by user intervention
 */
export type StoppedState = 'stopped'
```

Something to ponder: streams can be actually paused. 
So it's kind of a stream of events huh ?

**Comment by adrianhopebailie** [comment]
https://developer.mozilla.org/en-US/docs/Web/API/Streams_API/Concepts
```
The data is read sequentially in small pieces called chunks. A chunk can be a single byte, or it can be something larger such as a typed array of a certain size. A single stream can contain chunks of different sizes and types.
```

Seems a bit of a stretch to me having not used them before. Maybe someone with some prior experience with the stream API has a bit more imagination for it. 

> what this could look like using a ReadableStream

Is there a Duplex type stream? You could model all the paid packets/events/etc as a read only stream, but there's also the ability to adjust bandwidth. 

> don't know anything about STREAM 

I don't know much about what's under the hood either :)
I think this is a question for @sharafian / @adrianhopebailie

**Comment by adrianhopebailie** [comment]
@marcoscaceres 
Streams aside, what do you think about a `stopped` state and the new events as described above ?

**Comment by adrianhopebailie** [comment]
@adrianhopebailie we'd like to figure out how to implement this because currently the lack of a stopped state is impacting people who are using web monetization.

I wonder if we should be working on this repo only in the context of a longer term web standard and do our nearer term changes in the ILP RFCs?

**Comment by adrianhopebailie** [comment]
We should document everything here, not split the documentation.
(I will mark the ILP version as deprecated and point here, should have done that already!)

We have a spec (see index.html in this repo) and we can keep iterating on it as fast as we can experiment.

But, obviously, we should consider people who are already using the APIs so we should try and stay backwards compatible.

**Comment by adrianhopebailie** [comment]
This has been implemented in #11 and discussion related to adding a `ReadableStream` is in #27
